File: ruby/y2022/day02.rb
#!/usr/bin/env ruby
#
# Rock Paper Scissors

require_relative '../day'

class Day02 < Day
  SHAPE_SCORE = { rock: 1, paper: 2, scissors: 3 }
  WIN_SCORE = { lose: 0, draw: 3, win: 6 }
  # [theirs, yours] => win symbol
  MOVE_RESULT = {
    %i[rock rock] => :draw,
    %i[rock paper] => :win,
    %i[rock scissors] => :lose,
    %i[paper rock] => :lose,
    %i[paper paper] => :draw,
    %i[paper scissors] => :win,
    %i[scissors rock] => :win,
    %i[scissors paper] => :lose,
    %i[scissors scissors] => :draw
  }
  CODE_TO_SYM = {
    'A' => :rock,
    'B' => :paper,
    'C' => :scissors,
    'X' => :rock,
    'Y' => :paper,
    'Z' => :scissors
  }
  CODE_TO_RESULT = {
    'X' => :lose,
    'Y' => :draw,
    'Z' => :win
  }

  def part1
    do_part(:part1_score)
  end

  def part2
    do_part(:part2_score)
  end

  def do_part(score_method)
    lines = data_lines(1)
    score = 0
    lines.each do |line|
      round = line.split
      score += send(score_method, round)
    end
    puts score
  end

  def part1_score(round)
    round = round.map { |str| CODE_TO_SYM[str] }
    shape_score = SHAPE_SCORE[round[1]]
    move_result = MOVE_RESULT[round]
    win_score = WIN_SCORE[move_result]
    shape_score + win_score
  end

  def part2_score(round)
    theirs = CODE_TO_SYM[round[0]]
    desired_result = CODE_TO_RESULT[round[1]]
    win_entry = MOVE_RESULT.detect do |k, v|
      k[0] == theirs && v == desired_result
    end
    shape_score = SHAPE_SCORE[win_entry[0][1]]
    win_score = WIN_SCORE[desired_result]
    shape_score + win_score
  end
end

if __FILE__ == $PROGRAM_NAME
  require_relative '../aoc'

  aoc(2022, 2)
end

File: python/y2022/day02.py
# Rock Paper Scissors

from utils import *

SHAPE_SCORE = {"A": 1, "B": 2, "C": 3, "X": 1, "Y": 2, "Z": 3}
WIN_SCORE = {"lose": 0, "draw": 3, "win": 6}
# [theirs, yours] => win symbol
MOVE_RESULT = {
    ("A", "X"): "draw",
    ("A", "Y"): "win",
    ("A", "Z"): "lose",
    ("B", "X"): "lose",
    ("B", "Y"): "draw",
    ("B", "Z"): "win",
    ("C", "X"): "win",
    ("C", "Y"): "lose",
    ("C", "Z"): "draw",
}
CODE_TO_RESULT = {"X": "lose", "Y": "draw", "Z": "win"}
OPPONENT_AND_RESULT_TO_MY_SHAPE = {
    ("A", "win"): "Y",
    ("A", "lose"): "Z",
    ("A", "draw"): "A",
    ("B", "win"): "Z",
    ("B", "lose"): "X",
    ("B", "draw"): "Y",
    ("C", "win"): "X",
    ("C", "lose"): "Y",
    ("C", "draw"): "Z",
}


def part1(env):
    lines = data_file_lines(env)
    print(sum(part1_round_score(line) for line in lines))


def part2(env):
    lines = data_file_lines(env)
    print(sum(part2_round_score(line) for line in lines))


def part1_round_score(line):
    opponent_shape, my_shape = line.split()
    match MOVE_RESULT[(opponent_shape, my_shape)]:
        case "draw":
            return SHAPE_SCORE[my_shape] + 3
        case "win":
            return SHAPE_SCORE[my_shape] + 6
        case "lose":
            return SHAPE_SCORE[my_shape] + 0


def part2_round_score(line):
    opponent_shape, result_code = line.split()
    desired_result = CODE_TO_RESULT[result_code]
    my_shape = OPPONENT_AND_RESULT_TO_MY_SHAPE[(opponent_shape, desired_result)]
    match desired_result:
        case "draw":
            return SHAPE_SCORE[my_shape] + 3
        case "win":
            return SHAPE_SCORE[my_shape] + 6
        case "lose":
            return SHAPE_SCORE[my_shape] + 0
